### DISCUSSION TITLE
Connecting 2 Balun with Circuit

### DISCUSSION THREAD

**Discussion by PumadeGit** [scikit-rf/scikit-rf#931] [Q&A] [answered]
Hi, all,

I am new to scikit-rf. I just got an problem on making new circuits.
I would like to connect differential port of two 1:2 balun, to form a 2 port network.
```
                port 1-----o-----port 1
port 0---BALUN1                        BALUN2---port 0
                port 2-----o-----port 2
```
I tried a connection like this:
```
connexions = [
    [(balun1, 1), (balun2, 1)],
    [(balun2, 2), (balun1, 2)]
]
```
and connect it by,
`resulting_circuit = rf.Circuit(connexions)`

it just raised an error:
`AttributeError: A (network, port) node appears twice in the connection description.`
I have no ideas on this issue. I didn't see any port being connected twice.

Balun1 and 2 shared the same S3P file, Balun2 was obtained by balun1.copy().

Anyone can help on this issue? Thanks!

**Comment by jhillairet** [comment, marked as answer]
You need:
- do define the two Ports (cf [tutorial](https://scikit-rf.readthedocs.io/en/latest/tutorials/Circuit.html) for more details) and connect them
- to use unique names for each Network.So if blaun2 is a copy of balun1, the name should be changed 

```python
Port1 = rf.Circuit.Port(frequency=balun1.frequency, name='Port1')
Port2 = rf.Circuit.Port(frequency=balun1.frequency, name='Port2') # NB: frequency of balun1 and balun2 should be the same

balun2.name = 'balun2'

connexions = [
    [(port1, 0), (balun1, 0)],
    [(balun1, 1), (balun2, 1)],
    [(balun2, 2), (balun1, 2)],
    [(balun2, 0), (port2, 0)]
]
```

**Reply by PumadeGit** [reply]
Thanks! I missed the warning parts in Tutorial.

I followed your instruction, it worked!

By the way, the s3p file I used is from mini-circuit. As for these kind of multi port device, how do I get the port number definition of those networks (it's relationship to ports on physical device). There's some rules to follow? or can be get from some description?

Is there's anyway to check some detailed information of each port? 

Thanks!

**Reply by jhillairet** [reply]
> By the way, the s3p file I used is from mini-circuit. As for these kind of multi port device, how do I get the port number definition of those networks (it's relationship to ports on physical device). There's some rules to follow? or can be get from some description?

well, I suppose this should be indicated somehow on the vendor webpage...  datasheets usually also show some curves, so you can plot the S-parameters to deduce how the ports are ordered.
 
> Is there's anyway to check some detailed information of each port?

what kind of information?

**Reply by PumadeGit** [reply]
> > By the way, the s3p file I used is from mini-circuit. As for these kind of multi port device, how do I get the port number definition of those networks (it's relationship to ports on physical device). There's some rules to follow? or can be get from some description?
> 
> well, I suppose this should be indicated somehow on the vendor webpage...  datasheets usually also show some curves, so you can plot the S-parameters to deduce how the ports are ordered.

Thanks! It seems to be a practical way.
>  
> > Is there's anyway to check some detailed information of each port?
> 
> what kind of information?
> 
Description of the device (or say network), like summary method in Pandas Dataframe class, showing some meta description of the sNp file. Will that being useful?

**Reply by jhillairet** [reply]
> > what kind of information?
> 
> Description of the device (or say network), like summary method in Pandas Dataframe class, showing some meta description of the sNp file. Will that being useful?

It is already possible, for example by filling the parameters `comments` of a `Network`. These comments are automatically read when they exist in a Touchstone file... If they exist...

**Reply by PumadeGit** [reply]
Great to know! Thanks, I found their port definition in comments! There's a pin to port table.
Thanks!
